### DISCUSSION TITLE
4.0.0-beta.3 :: Global Styles :: Recommended solution for using tokens in global classes?

### DISCUSSION THREAD

**Discussion by mcmimik** [primefaces/community#2036] [PrimeVue]
Still thrilled with the theming in version 4. However, we're facing an issue: we can't figure out how to pass tokens into custom global styles.

The task is simple: we want to have global classes with our own prefix. Currently, we use CSS Modules for this, adding the prefix through the Vite generateScopedName function. But inside the modules, we use CSS variables (corresponding to tokens), and we **can't add prefixes** to them.

For example, we have a main `global.module.css` module:
```css
@import "vars.css";
@import "typography.css";
```

In the `vars.css` module:
```css
@layer vars {
  :root {
    /* ... */
    --t2-fs: 14px;
    --t2-lh: 16px;
    /* ... */
  }
}
```

In the `typography.css` file, we have a class like this:
```css
@layer typography {
  .t2 {
    font-size: var(--t2-fs);
    line-height: var(--t2-lh);
  }
}
```

And in the `paginator.module.css` module:
```css
@layer paginator {
  .current {
    composes: t2 from "@/styles/typography.module.css";
  }
}
```

Finally, we pass the CSS module into `pt`:
```js
import uiCss from "@/styles/ui.module.css";
import paginator from "@/components/paginator/paginator.module.css";

app.use(PrimeVue, {
  css: uiCss,
  pt: {
    paginator,
  },
});
```

And we get the following CSS:
```css
.prefix-t2 {
  font-size: var(--t2-fs);
  line-height: var(--t2-lh);
}
```

The problem is that the CSS variables remain _unprefixed_, and we really need them to have a prefix.

Currently, our tokens are declared in the theme file within the `primitive` object.
Is there a way to define **global classes that use the declared tokens from the theme**?

For now, a workaround seems to be:

1. Create a `global.js` file with global classes as functions:
```js
import { $dt } from "primevue/themes";

export const t1 = () => ({
  fontSize: $dt("text.t1.font.size").value,
  lineHeight: $dt("text.t1.line.height").value,
});

export const t2 = () => ({
  fontSize: $dt("text.t2.font.size").value,
  lineHeight: $dt("text.t2.line.height").value,
});
```

2. Destructure the global classes into the `style` attribute through the element function within the `pt` object. Thus, the component styling looks something like this:
```js
import { $dt } from "primevue/themes";
import paginator from "@/components/paginator/paginator.module.css";
import { t2 } from "@/styles/global";

app.use(PrimeVue, {
  css: uiCss,
  pt: {
    paginator: {
      ...paginator,
      current: () => ({
        class: paginator.current,
        style: {
          marginLeft: $dt("size.x2").value,
          ...t2(),
        },
      }),
    },
  },
});
```

It works, but seems to be too complicated.

Ideally, we would like to be able to somehow **reference tokens from CSS modules**, leaving the `pt` object only for "states" (`active`, etc.).

Any hints or recommendations on how to implement global classes with tokens would be greatly appreciated.

**Comment by cagataycivici** [comment]
Thanks for the feedback, we'll discuss this as an enhancement. Really interested to see how the new theming would be utilized and extended.

**Reply by mcmimik** [reply]
Thank you for considering this enhancement. Here are a few scenarios where creating global styles tied to tokens with prefixed CSS variables could be incredibly useful:

1. **Avoiding conflicts with multiple UIKit versions:** When multiple versions of a UIKit are used on the same page (such as in micro-frontends), having prefixed CSS variables ensures that styles do not conflict with each other. We use the version from package.json in the prefix, which helps reuse the same versions of the library while making them independent of each other.
2. **Simplifying debugging:** Prefixes in CSS variables and global classes make debugging a breeze, especially in enterprise environments. When users report issues with just a screenshot of the developer tools, the prefixed variables make it extremely easy to identify and address the problem quickly.
3. **Creating custom Utility classes:** This allows developers to create their own utility class systems that are seamlessly integrated with the token system. It is a really nice alternative to third-party CSS frameworks, especially when you need a fairly limited set of own classes.
